### ISSUE TITLE
use on-demand choice constraints for handling coercions

### ISSUE THREAD

**Issue by leodemoura** [leanprover/lean3#252]
Here is a problematic example:

```
open nat

inductive tree (A : Type) :=
leaf : A → tree A,
node : tree A → tree A → tree A

check tree.node

definition size {A : Type} (t : tree A) :=
tree.rec (λ a, 1) (λ t₁ t₂ n₁ n₂, n₁ + n₂) t
--            ^^^^ problem is here
check size

eval size (tree.node (tree.node (tree.leaf 0) (tree.leaf 1))
                (tree.leaf 0))
```

**Comment by leodemoura** [comment]
@fpvandoorn 
Lean is now automatically lifting coercions from `A` to `B` into coercions from `D -> A` to `D -> B`.
It also lifts them when performing the backtracking search. 
This feature solves the problem above, and also fixed #644.
This feature generated problems in two files in the HoTT library. I fixed them by using `proof ... qed`.
If you have problems, you may also use the new option

``` lean
set_option elaborator.lift_coercions false
```

**Comment by fpvandoorn** [comment]
Ok, thanks for the option. I'll try to avoid it by also using `proof ... qed`.
